kratos/mpi/sources/mpi_data_communicator.cpp

```diff
@@ -474,6 +474,18 @@ bool MPIDataCommunicator::IsNullOnThisRank() const
     return mComm == MPI_COMM_NULL;
 }
 
+const DataCommunicator& MPIDataCommunicator::GetSubDataCommunicator(
+    const std::vector<int>& rRanks,
+    const std::string& rNewCommunicatorName
+    ) const
+{
+    if (ParallelEnvironment::HasDataCommunicator(rNewCommunicatorName)) {
+        return ParallelEnvironment::GetDataCommunicator(rNewCommunicatorName);
```

Comment:
this is problematic because the ranks passed here could be different!

looking at this again I am leaning to make the creation explicit :thinking: 

other opinions? This is important in terms of the concept

Reply:
I understand your point...

Reply:
Would fix the problem this?

~~~C++
const DataCommunicator& MPIDataCommunicator::GetSubDataCommunicator(
    const std::vector<int>& rRanks,
    const std::string& rNewCommunicatorName
    ) const
{
    int has_data_communicator = ParallelEnvironment::HasDataCommunicator(rNewCommunicatorName);
    has_data_communicator = SumAll(has_data_communicator);
    KRATOS_ERROR_IF_NOT(has_data_communicator == 0 || has_data_communicator == Size()) << "The communicator " << rNewCommunicatorName << " already exists in some partitions but not in some others." << std::endl;
    if (has_data_communicator) {
        return ParallelEnvironment::GetDataCommunicator(rNewCommunicatorName);
    } else {
        return DataCommunicatorFactory::CreateFromRanksAndRegister(*this, rRanks, rNewCommunicatorName);
    }
}
~~~

Reply:
or:

~~~cpp
const DataCommunicator& MPIDataCommunicator::GetSubDataCommunicator(
    const std::vector<int>& rRanks,
    const std::string& rNewCommunicatorName
    ) const
{
    if (ParallelEnvironment::HasDataCommunicator(rNewCommunicatorName)) {
        const DataCommunicator& r_data_communicator = ParallelEnvironment::GetDataCommunicator(rNewCommunicatorName);
        const int rank = Rank();
        const auto it_find = std::find(rRanks.begin(), rRanks.end(), rank);
        if (it_find != rRanks.end()) {
            KRATOS_ERROR_IF_NOT(IsDefinedOnThisRank()) << "The rank " << rank << " is not defined in the existing data communicator " << rNewCommunicatorName  << " despite being in the provided rank list" << std::endl;
        } else {
            KRATOS_ERROR_IF_NOT(IsNullOnThisRank()) << "The rank " << rank << " is defined in the existing data communicator " << rNewCommunicatorName << " despite not being in the provided rank list" << std::endl;
        }
        return r_data_communicator;
    } else {
        return DataCommunicatorFactory::CreateFromRanksAndRegister(*this, rRanks, rNewCommunicatorName);
    }
}
~~~